azkaban-common/src/main/java/azkaban/project/AzkabanProjectLoader.java

```diff
@@ -227,6 +222,19 @@ private void persistProject(final Project project, final FlowLoader loader, fina
     }
   }
 
+  private void uploadFlowFilesRecursively(final File projectDir, final Project project, final int
+      newProjectVersion) {
+    for (final File file : projectDir.listFiles(new SuffixFilter(Constants.FLOW_FILE_SUFFIX))) {
```

Comment:
Is two for-loop a duplicate? if me, I would use one for loop to walk through all the potential files.

Reply:
One for-loop is to traverse .flow files in the current directory, the other for loop is to traverse the subdirectories.

Reply:
The filter can only be applied to directories, not individual files. If we want to use the existing filter api, then we cannot avoid two for-loops. Otherwise, we need to list all the files first, and then implement our own filtering logic on each individual file inside the directory, which I think is not quite necessary.